### ISSUE TITLE
How to use i2c as slave without directly control the registries 

### ISSUE THREAD

**Issue by eisterman** [stm32-rs/stm32f1xx-hal#163]
Is there an official method from this crate to use i2c without directly manipulate the bit of the i2c registries?
And if not, why?

**Comment by TeXitoi** [comment]
No, because no one have done it yet.

Notably, it lacks an I2C slave embedded trait in embedded hal.

Contribution welcome!

**Comment by eisterman** [comment]
If there isn't any problem with it I'm glad to implement it

**Comment by TeXitoi** [comment]
Feel free to propose a first POC and ask for feedbacks. You can also chat with a lot of rust embedded developers on matrix https://github.com/rust-embedded/wg#contact

**Comment by eisterman** [comment]
I ask to confirm my doubt.
Is the non-blocking master i2c interface currently lacking any method to be used, unlike its blocking counterpart, correct?

**Comment by TeXitoi** [comment]
I'm not sure to understand. Doesn't the slave interface by definition different because of responding to requests?

**Comment by eisterman** [comment]
Yes, however, there are two versions of the master, one blocker and one no. However, I noticed that only the blocking version is public (thanks to the embedded_hal traits). I was wondering if this was a lack of the current implementation or had a reason to reflect it in the slave implementation.

In general, I am thinking of managing the slave in a similar way to how the Arduino Wire library does, or through the assignment of two closures, one for sending data and one for receiving data.
In this area, many recommendations are accepted!

**Comment by therealprof** [comment]
I think you're talking across one another.

Indeed there's no I2C slave traits to be implemented and there's also no custom I2C slave implementation implemented in this crate.

It would be great to have both. 

CC https://github.com/rust-embedded/embedded-hal/issues/179

**Comment by TheZoq2** [comment]
> Yes, however, there are two versions of the master, one blocker and one no. However, I noticed that only the blocking version is public (thanks to the embedded_hal traits). I was wondering if this was a lack of the current implementation or had a reason to reflect it in the slave implementation.

Both versions are public, but only blocking i2c uses embedded hal traits, since there is no hal traits for non-blocking i2c. The same is true for slave i2c, so if you do implement it right now, users have to depend on stm32f1xx_hal directly.

The alternative is to add a slave trait to embedded_hal, and then implement it here. That would obviously be more work but is a nicer implementation.

**Comment by ostenning** [comment]
Wondering if there was any progress for an I2C slave feature in this crate?

Would a PR be welcome for implementing this as a device specific (ignoring the embedded-hal traits) feature for now?

**Comment by burrbull** [comment]
> Would a PR be welcome for implementing this as a device specific (ignoring the embedded-hal traits) feature for now?

Sure.

**Comment by surban** [comment]
https://github.com/surban/openemc/blob/master/openemc-firmware/src/i2c_slave.rs

**Comment by ostenning** [comment]
> https://github.com/surban/openemc/blob/master/openemc-firmware/src/i2c_slave.rs
> 
> and maybe https://github.com/surban/openemc/blob/master/openemc-firmware/src/i2c_reg_slave.rs
> 
> Let me know if this is useful. If so, I will open a PR.

That's quite helpful, thank you

**Comment by phansel** [comment]
backlink: https://github.com/rust-embedded/embedded-hal/issues/179#issuecomment-4188310685

> Some examples of I2C peripheral/slave mode have been built for rp2040:
> https://github.com/embassy-rs/embassy/blob/main/examples/rp/src/bin/i2c_slave.rs
